lib/util/github/graphql/datasource-helper.ts

```diff
@@ -222,14 +226,27 @@ export class GithubGraphqlDatasourceHelper<
     return res;
   }
 
-  private async doPaginatedQuery(): Promise<ResultItem[]> {
-    const resultItems: ResultItem[] = [];
+  private cacheAdapter: GithubGraphqlCacheAdapter<ResultItem> | undefined;
+  private getCacheAdapter(): GithubGraphqlCacheAdapter<ResultItem> {
+    if (this.cacheAdapter) {
+      return this.cacheAdapter;
+    }
+    const cacheNs = this.getCacheNs();
+    const cacheKey = this.getCacheKey();
+    this.cacheAdapter = this.isCacheable
```

Comment:
do we instantiate this class for every package name?

Reply:
Yes, it encapsulates the complex state of pagination logic that otherwise would be handled as many local variables declared in one big function.

Reply:
To be more specific, we have one complex asynchronous loop that handles things like:
- Pagination
- Shrink and retry
- Upper limit for queries to be made
- Private/public detection
- Cache load/update/store
- Based on cached content, determine whether we need more pages or not

Cache adapters meant to handle last two points of this list.